File: modules/logparser/drain_log_parser.py
# SPDX-License-Identifier: MIT
# This file implements the Drain algorithm for log parsing.
# Based on https://github.com/logpai/logparser/blob/master/logparser/Drain/Drain.py by LogPAI team


from itertools import chain

from .base_log_parser import Token
from .drain_base_log_parser import Content, DrainBaseLogParser, LogCluster, Node
from .parser_factory import parser_register


@parser_register
class DrainLogParser(DrainBaseLogParser):
    @staticmethod
    def _create_template(content1: Content, content2: Content) -> Content:
        assert len(content1) == len(content2)
        return [
            token1 if token1 == token2 else "<*>"
            for token1, token2 in zip(content1, content2)
        ]

    def _add_seq_to_prefix_tree(self, cluster: LogCluster) -> None:
        content_length = len(cluster.content)
        cur_node = self._root_node
        for cur_node_depth, token in enumerate(
            chain([Token(content_length)], cluster.content),
            start=1,
        ):
            # if at max depth or this is last token in template - add current log cluster to the leaf node
            if cur_node_depth == self._depth or cur_node_depth == content_length + 2:
                cur_node.cluster_ids.append(cluster.cluster_id)
                return

            # if token not matched in this layer of existing tree.
            if token not in cur_node.children_node:
                if len(cur_node.children_node) + 1 < self._children:
                    # 如果当前节点不是最后一个节点，就添加一个新的节点
                    new_node = Node()
                    cur_node.children_node[token] = new_node
                    cur_node = new_node
                elif len(cur_node.children_node) + 1 == self._children:
                    # 如果当前节点是最后一个节点，就添加一个新的通配符节点
                    # 注意由于 _parametrize_numeric_tokens 的存在，可能已经存在一个通配符节点了，所以需要先检查一下
                    new_node = Node()
                    cur_node.children_node["<*>"] = new_node
                    cur_node = new_node
                else:
                    # 如果当前节点已满，就直接使用通配符节点
                    cur_node = cur_node.children_node["<*>"]

            # if the token is matched
            else:
                cur_node = cur_node.children_node[token]

    @staticmethod
    def _get_seq_distance(
        content1: Content,
        content2: Content,
        include_params: bool,
    ) -> tuple[float, int]:
        # content1 is a template, content2 is the log to match
        assert len(content1) == len(content2)

        # list are empty - full match
        if len(content1) == 0:
            return 1.0, 0

        sim_tokens = 0
        param_count = 0

        for token1, token2 in zip(content1, content2):
            if token1 == "<*>":
                param_count += 1
                continue
            if token1 == token2:
                sim_tokens += 1

        if include_params:
            # 参数位也当匹配贡献
            sim_tokens += param_count

        return float(sim_tokens) / len(content1), param_count

    def _tree_search(self, content: Content, include_params: bool) -> LogCluster | None:
        content_length = len(content)
        cur_node = self._root_node

        for cur_node_depth, token in enumerate(
            chain([Token(content_length)], content),
            start=1,
        ):
            # at max depth or this is last token
            if cur_node_depth == self._depth or cur_node_depth == content_length + 2:
                break

            if token in cur_node.children_node:
                cur_node = cur_node.children_node[token]
            elif (
                "<*>" in cur_node.children_node
            ):  # no exact next token exist, try wildcard node
                cur_node = cur_node.children_node["<*>"]
            else:  # no wildcard node exist
                return None

        # get best match among all clusters with same prefix, or None if no match is above sim_th
        return self._fast_match(cur_node.cluster_ids, content, include_params)

    @staticmethod
    def name() -> str:
        return "Drain"

    @staticmethod
    def description() -> str:
        return "Drain 是一种基于树结构的高效日志模板提取算法。"

File: modules/logparser/drain_log_parser.cxx
#include "drain_log_parser.hxx"
#include "utils.hxx"
#include <ranges>

namespace logparser
{

DrainLogParser::DrainLogParser(uint16_t depth, uint16_t children, float sim_thr):
    m_depth(depth), m_children(children), m_sim_thr(sim_thr)
{}

// std::vector<std::string> DrainLogParser::parse(const std::vector<TContent>& logs)
// {
//     // Initialize the prefix tree
//     this->m_root = std::make_unique<Node>();
//     this->m_cluster_pool.clear();

//     std::vector<LogCluster*> cluster_results;
//     cluster_results.reserve(logs.size());

//     for (const auto& log : logs)
//     {
//         cluster_results.push_back(this->_add_content(log));
//     }

//     std::vector<std::string> templates;
//     templates.reserve(logs.size());

//     for (const auto& cluster : cluster_results)
//     {
//         templates.push_back(cluster->get_template());
//     }

//     return templates;
// }

void DrainLogParser::parse(const std::string& log_file, const std::string& log_format, const std::vector<std::string>& group_names)
{
    auto relation = load_data(log_file, log_format, group_names);
}

DrainLogParser::LogCluster* DrainLogParser::_add_content(const TContent& content)
{
    auto match_cluster {this->_tree_search(content)};

    // Match no existing log cluster
    if (!match_cluster)
    {
        this->m_cluster_pool.emplace_back(content);
        match_cluster = &this->m_cluster_pool.back();
        this->_add_to_prefix_tree(match_cluster);
    }
    else
    {
        // Add the new log message to the existing cluster
        auto new_content {DrainLogParser::_create_template(content, match_cluster->content)};
        if (new_content != match_cluster->content)
        {
            match_cluster->content = new_content;
        }
    }

    return match_cluster;
}

DrainLogParser::LogCluster* DrainLogParser::_tree_search(const TContent& content, bool include_params)
{
    auto length {content.size()};
    auto length_token = std::to_string(length);
    auto cur_node {this->m_root.get()};

    // for (const auto& [i, token] : std::views::enumerate(std::views::concat(std::views::single(length_token), content)))
    // {
    //     auto cur_node_depth {static_cast<size_t>(i) + 1};

    //     // at max depth or this is last token
    //     if (cur_node_depth == this->m_depth || cur_node_depth == length + 2)
    //     {
    //         // get best match among all clusters with same prefix, or None if no match is above sim_th
    //         break;
    //     }

    //     if (auto it {cur_node->children_node.find(token)}; it == cur_node->children_node.end())
    //     {
    //         it = cur_node->children_node.find(WILDCARD);
    //         if (it == cur_node->children_node.end())
    //         {
    //             // no wildcard node exist
    //             return nullptr;
    //         }
    //         cur_node = it->second.get();
    //     }
    //     else
    //     {
    //         cur_node = it->second.get();
    //     }
    // }

    return this->_fast_match(content, cur_node, include_params);
}

DrainLogParser::LogCluster* DrainLogParser::_fast_match(const TContent& content, const Node* node, bool include_params)
{
    float       max_sim {-1.0F};
    uint16_t    max_param_count {0};
    LogCluster* max_cluster {};

    for (const auto& cluster : node->clusters)
    {
        auto [cur_sim, param_count] {this->_get_distance(content, cluster->content, include_params)};
        if (cur_sim > max_sim || (cur_sim == max_sim && param_count > max_param_count))
        {
            max_sim         = cur_sim;
            max_param_count = param_count;
            max_cluster     = cluster;
        }
    }

    if (max_sim > this->m_sim_thr)
    {
        return max_cluster;
    }

    return nullptr;
}

void DrainLogParser::_add_to_prefix_tree(LogCluster* cluster)
{
    auto length {cluster->content.size()};
    auto length_token = std::to_string(length);
    auto cur_node {this->m_root.get()};

    // for (const auto& [i, token] : std::views::enumerate(std::views::concat(std::views::single(length_token), cluster->content)))
    // {
    //     auto cur_node_depth {static_cast<size_t>(i) + 1};

    //     // if at max depth or this is last token in template - add current log cluster to the leaf node
    //     if (cur_node_depth == this->m_depth || cur_node_depth == length + 2)
    //     {
    //         // at max depth or this is last token
    //         cur_node->clusters.push_back(cluster);
    //         return;
    //     }

    //     if (auto it {cur_node->children_node.find(token)}; it == cur_node->children_node.end())
    //     {
    //         // if token not matched in this layer of existing tree.
    //         if (cur_node->children_node.size() + 1 < this->m_children)
    //         {
    //             // 如果当前节点不是最后一个节点，就添加一个新的节点
    //             auto [it, _] = cur_node->children_node.emplace(token, std::make_unique<Node>());
    //             cur_node     = it->second.get();
    //         }
    //         else if (cur_node->children_node.size() + 1 == this->m_children)
    //         {
    //             // 如果当前节点是最后一个节点，就添加一个新的通配符节点
    //             auto [it, _] = cur_node->children_node.emplace(WILDCARD, std::make_unique<Node>());
    //             cur_node     = it->second.get();
    //         }
    //         else
    //         {
    //             // 如果当前节点已满，就直接使用通配符节点
    //             cur_node = cur_node->children_node[WILDCARD].get();
    //         }
    //     }
    //     else
    //     {
    //         // if the token is matched
    //         cur_node = it->second.get();
    //     }
    // }
}

std::pair<float, uint16_t> DrainLogParser::_get_distance(const TContent& content1, const TContent& content2, bool include_params)
{
    // list are empty - full match
    if (content1.size() == 0)
    {
        return {1.0F, 0};
    }

    uint16_t sim_tokens {0};
    uint16_t param_count {0};

    for (const auto& [token1, token2] : std::views::zip(content1, content2))
    {
        // 这里的 content2 是模板，所以这里要用 token2 来判断是否是参数位
        if (token2 == WILDCARD)
        {
            ++param_count;
            continue;
        }

        if (token1 == token2)
        {
            ++sim_tokens;
        }
    }

    if (include_params)
    {
        // 参数位也当匹配贡献
        sim_tokens += param_count;
    }

    return {static_cast<float>(sim_tokens) / content1.size(), param_count};
}

TContent DrainLogParser::_create_template(const TContent& content1, const TContent& content2)
{
    TContent new_content {};
    new_content.reserve(content1.size());

    for (const auto& [token1, token2] : std::views::zip(content1, content2))
    {
        if (token1 == token2)
        {
            new_content.push_back(token1);
        }
        else
        {
            new_content.push_back(WILDCARD);
        }
    }

    return new_content;
}

std::string DrainLogParser::LogCluster::get_template() const
{
    auto view = this->content | std::views::join_with(' ');
    return std::string(view.begin(), view.end());
}

}    // namespace logparser